stablehlo/dialect/AssemblyFormat.cpp

```diff
@@ -252,6 +602,63 @@ ParseResult parseSelectOpType(OpAsmParser& parser, Type& pred, Type& onTrue,
                                 fnType);
 }
 
+void printWhileOp(OpAsmPrinter& p, Operation* op, Region& cond, Region& body) {
+  p << '(';
+  llvm::interleaveComma(llvm::zip(body.getArguments(), op->getOperands()), p,
+                        [&](auto zip) {
+                          p.printOperand(std::get<0>(zip));
+                          p << " = ";
+                          p.printOperand(std::get<1>(zip));
+                        });
+  p << ")";
+  if (op->getNumOperands()) {
+    p << " : ";
+    llvm::interleaveComma(op->getOperandTypes(), p);
+  }
+  p.printOptionalAttrDictWithKeyword(op->getAttrs());
+  p.printNewline();
+  p << " cond ";
+  p.printRegion(cond, /*printEntryBlockArgs=*/false);
+  p << " do ";
+  p.printRegion(body, /*printEntryBlockArgs=*/false);
+}
+
+ParseResult parseWhileOp(OpAsmParser& parser, OperationState& result) {
+  llvm::SMLoc loc = parser.getCurrentLocation();
+  // Parse the operands of the while: these are of the form:
+  //   %iter_arg = %init_val
+  // where %iter_arg is the name of the block argument in the cond/body blocks
+  // and %init_val is the actual operand.
+  SmallVector<OpAsmParser::UnresolvedOperand> operands;
+  SmallVector<OpAsmParser::UnresolvedOperand> iterArgs;
+  if (parser.parseLParen()) return failure();
+  do {
+    if (succeeded(parser.parseOptionalRParen())) break;
+    OpAsmParser::UnresolvedOperand operand, iterArg;
+    if (parser.parseOperand(iterArg) || parser.parseEqual() ||
+        parser.parseOperand(operand))
+      return failure();
+    iterArgs.push_back(iterArg);
+    operands.push_back(operand);
+    if (succeeded(parser.parseOptionalRParen())) break;
+    if (failed(parser.parseComma())) return failure();
+  } while (true);
```

Comment:
Consider adding a newline after this `while (true);` to help with reading.

Reply:
refactored to while loop instead of do/while